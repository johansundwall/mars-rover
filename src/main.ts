import { getBoundaries } from './plateau/getBoundaries';
import { createRovers } from './rover/createRovers';
import { executeInstructions } from './rover/executeInstructions';
import { Rover } from './rover/rover.types';
import { generateOutput } from './utils/generateOutput';

export const main = (input) => {
  const rovers = createRovers(input);
  const upperBoundaries = getBoundaries(input);

  const occupiedPositions = [];
  const crashedRovers = [];
  const finishedRovers = [];

  rovers.forEach((rover: Rover, index: number) => {
    try {
      const finishedRover = executeInstructions({
        rover, occupiedPositions, upperBoundaries, roverNumber: index,
      });
      occupiedPositions.push({ x: finishedRover.x, y: finishedRover.y });
      finishedRovers.push(finishedRover);
    } catch (err) {
      crashedRovers.push({ error: err, rover, index });
    }
  });

  if (crashedRovers.length) {
    console.error(crashedRovers); // iterate over and make more ready friendly
  }

  return generateOutput(finishedRovers);
};
