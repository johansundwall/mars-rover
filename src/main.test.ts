import { main } from './main';

describe('main function', () => {
  it('should match snapshot', () => {
    const input = `5 5
    1 2 N
    LMLMLMLMM
    3 3 E
    MMRMMRMRRM
    `;

    expect(main(input)).toMatchSnapshot();
  });

  it('should match snapshot when there has been colissions and call console.error', () => {
    global.console = {
      ...global.console,
      error: jest.fn(),
    };

    const input = `5 5
    1 2 N
    LMLMLMLMM
    1 2 N
    LMLMLMLMM
    `;

    expect(main(input)).toMatchSnapshot();
    expect(console.error).toHaveBeenCalled();
  });
});
