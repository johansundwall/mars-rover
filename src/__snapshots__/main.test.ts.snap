// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`main function should match snapshot 1`] = `
"1 3 N
5 1 E
"
`;

exports[`main function should match snapshot when there has been colissions and call console.error 1`] = `
"1 3 N
"
`;
